core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -196,54 +196,44 @@ protected Object handleReadCommand(InvocationContext ctx, DataCommand command) t
       if (trace) {
-         log.tracef("Locally prefetching entry for key %s", key);
+         log.tracef("Locally prefetched entry %s", entry);
       }
-      GetCacheEntryCommand getCacheEntryCommand = commandsFactory.buildGetCacheEntryCommand(key, segment, EnumUtil.bitSetOf(Flag.CACHE_MODE_LOCAL));
-      getCacheEntryCommand.setTopologyId(command.getTopologyId());
-      return makeStage(invokeNextThenApply(ctx, getCacheEntryCommand, handleLocallyLookedUpEntry));
-   }
-
-   private Object handleLocallyLookedUpEntry(InvocationContext ctx1, VisitableCommand command1, Object rv) {
-      GetCacheEntryCommand cmd = (GetCacheEntryCommand) command1;
-      if (trace) {
-         log.tracef("Locally prefetched entry %s", rv);
-      }
-      CacheEntry entry = (CacheEntry) rv;
       Metadata metadata = entry != null ? entry.getMetadata() : null;
       CompletionStage<InternalCacheValue> future;
-      int segment = cmd.getSegment();
       if (metadata != null && metadata.version() != null && svm.isVersionActual(segment, metadata.version())) {
          return null;
       } else if ((metadata instanceof RemoteMetadata) &&
             svm.getSegmentState(segment) == ScatteredVersionManager.SegmentState.VALUE_TRANSFER) {
          // The RemoteMetadata is valid only during value transfer - in blocked state there shouldn't be any such
          // entry and during key transfer we could see metadata pointing to a node with outdated information.
          Address backup = ((RemoteMetadata) metadata).getAddress();
-         future = retrieveRemoteValue(Collections.singleton(backup), cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(Collections.singleton(backup), key, segment, cmd);
       } else {
-         future = retrieveRemoteValue(null, cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(null, key, segment, cmd);
       }
       return asyncValue(future.thenAccept(maxValue -> {
          if (maxValue == null) {
             return;
          }
          // The put below could fail updating the context if the data container got the updated value while we were
-         // prefetching that. Also, we need to call RepeatableReadEntry.updatePreviousValue() to get return value
+         // prefetching that (then the version would not be higher than the on in DC).
+         // We need to call RepeatableReadEntry.updatePreviousValue() (through wrapExternalEntry) to get return value
          // from the main command correct.
-         entryFactory.wrapExternalEntry(ctx1, cmd.getKey(), maxValue.toInternalCacheEntry(cmd.getKey()), true, true);
-         PutKeyValueCommand putKeyValueCommand =
-               commandsFactory.buildPutKeyValueCommand(cmd.getKey(), maxValue.getValue(), segment, maxValue.getMetadata(),
-                                                       STATE_TRANSFER_FLAGS);
+         entryFactory.wrapExternalEntry(ctx, key, maxValue.toInternalCacheEntry(key), true, true);
```

Comment:
Sorry Radim, but every time you add more info it raises more questions for me :(

* Who checks that the version in the PKVC is higher than the version in the DC?
* What happens if the version in DC increases after `PrefetchInterceptor`?
* This code runs before EWI. So why do we have to worry about the previous value specifically, do we expect the context to already have a previous value for this key that's wrong?

Reply:
1) ScatteredDistributionInterceptor
2) The value from put-for-state-transfer is not written in SDI. It is not possible that the version in DC increases in other way that receiving the most recent value, because there is always only one node that can write it and assign a version number (this is synced using `CONFIRM_REVOKED SEGMENTS`).
3) EWI cannot read the value because there's no read owner at this point. `PrefetchInterceptor` retrieves the value from all nodes and picks the highest version, or uses `RemoteMetadata` to retrieve it from single node if the location of the highest version is already known.